File: ngxl.rb
#!/usr/bin/env ruby
# encoding: UTF-8

require 'time'

class MinimumAverageMaximum
  attr_reader :key, :count, :min, :max, :total

  def initialize
    @count = 0.0

    @total = 0.0
    @min   = 0.0
    @max   = 0.0
  end

  def add(value)
    value = value.to_f

    if @count == 0.0
      @min = value
      @max = value
    else
      @min = value if value < @min
      @max = value if value > @max
    end

    @count += 1.0
    @total += value
  end

  def average
    @total / @count
  end
end

class Statuses
  attr_reader :key, :count, :code2, :code3, :code4, :code5

  def initialize
    @count = 0

    @code2 = 0
    @code3 = 0
    @code4 = 0
    @code5 = 0
  end

  def add(code)
    @count += 1

    case code[0..0]
    when '2'
      @code2 += 1
    when '3'
      @code3 += 1
    when '4'
      @code4 += 1
    when '5'
      @code5 += 1
    end
  end
end

REPORT = '--report'.freeze
BY     = '--by'.freeze
CSV    = '--csv'.freeze

REPORT_VALUES = {
  'status'   => { 'offset' => 1, 'class' => Statuses },
  'size'     => { 'offset' => 2, 'class' => MinimumAverageMaximum },
  'response' => { 'offset' => 3, 'class' => MinimumAverageMaximum }
}.freeze

BY_VALUES = %w[day hour ip path].freeze

def usage(message)
  puts "ERROR: #{message}"
  puts
  puts <<-EOM
ngxl --report status|size|reponse --by hour|day|ip|path [--csv] <list of Nginx log files>

  Will process the nginx log files and report either the status (by class),
  size of the response in bytes (giving the total, minimum, average and maximum)
  or response times (also with minimum, average and maximum)

  Each line of the report will be either the day or hour that the data was
  recorded in or the ip address that the request was made from

  The log lines can be read from stdin if you need to pipe them (through
  grep perhaps)

  The output will be formatted as CSV data if the "--csv" flag is given
EOM

  exit
end

def opts(flags)
  ##
  # Assumes that the options, in long format such as "--file xxx",
  # will come before the arguments. The first argument that appears
  # will signal the end of the options. So with:
  #
  # --file fred.txt --delim tab arnold smith --output fred.out
  #
  # "--file" and "--delim" are options and "arnold", "smith",
  # "--output", "fred.out" are all arguments
  #
  # To force option like arguments from being read as options
  # the plain "--" option will stop looking for options. Everything
  # after the "--" will be arguments
  #
  # All options must have parameters
  #
  # Probably a tad over-engineered :)
  ##

  commands = ARGV.dup

  options = {}
  arguments = []

  end_of_options = false

  while commands.any?
    x = commands.shift.downcase

    if end_of_options
      arguments << x
    elsif x == '--'
      end_of_options = true
    elsif x.index('--') == 0
      if flags.include?(x)
        y = true
      else
        y = commands.shift.downcase
      end
      options[x] = y
    else
      arguments << x
      end_of_options = true
    end
  end

  [options, arguments]
end

def valid_options(options)
  usage "Missing option #{REPORT}" unless options.key?(REPORT)
  usage "Argument to #{REPORT} must be one of '#{REPORT_VALUES.keys.join("', '")}'" unless REPORT_VALUES.key?(options[REPORT])

  usage "Missing option #{BY}" unless options.key?(BY)
  usage "Argument to #{BY} must be one of '#{BY_VALUES.keys.join("', '")}'" unless BY_VALUES.include?(options[BY])

  x = options.keys
  x.delete(REPORT)
  x.delete(BY)
  x.delete(CSV)

  return if x.size == 0

  usage "Unknown option(s) #{x.inspect}"
end

def process_line_from_file(fh, data, by, report_values)
  fh.each do |line|
    parts = line.split(/\s+/)
    pos = nil
    i = 0

    parts.each do |e|
      if e.include?('HTTP/1')
        pos = i
        break
      end
      i += 1
    end

    next unless pos

    k = case by
        when 'ip'
          parts[0]
        when 'day'
          parts[pos - 4][1..11]
        when 'hour'
          parts[pos - 4][1..14]
        when 'path'
          parts[pos - 1].split('?').first
        end

    data[k] = report_values['class'].new unless data.key?(k)
    data[k].add(parts[pos + report_values['offset']])
  end
end

def format_date(date)
  nk = date.dup
  nk[11] = ' '
  nk += ':00:00'

  Time.parse(nk).strftime('%Y-%m-%d %H')
rescue
end

def format_day(date)
  nk = date.dup
  nk += ' 00:00:00'

  Time.parse(nk).strftime('%Y-%m-%d')
rescue
  nk
end

def setup_display(by, report, data)
  header = []
  line = []

  case by
  when 'hour'
    header << '%-13s' % 'date_and_hour'
    line << '%-13s'
  when 'day'
    header << '%-10s' % 'day'
    line << '%-10s'
  when 'ip'
    header << '%-15s' % 'ip_address'
    line << '%-15s'
  when 'path'
    size = 15
    data.keys.each do |k|
      size = k.size if k.size > size
    end

    header << "%-#{size}s" % 'path'
    line << "%-#{size}s"
  end

  case report
  when 'status'
    header << '%8s' % 'count'
    header << '%6s' % '2xx'
    header << '%6s' % '3xx'
    header << '%6s' % '4xx'
    header << '%6s' % '5xx'

    line += ['%8d', '%6d', '%6d', '%6d', '%6d']
  when 'size'
    header << '%8s' % 'count'
    header << '%15s' % 'total'
    header << '%15s' % 'min'
    header << '%15s' % 'avg'
    header << '%15s' % 'max'

    line += ['%8d', '%15d', '%15d', '%15d', '%15d']
  when 'response'
    header << '%8s' % 'count'
    header << '%8s' % 'min'
    header << '%8s' % 'avg'
    header << '%8s' % 'max'

    line += ['%8d', '%8.3f', '%8.3f', '%8.3f']
  end

  [header, line]
end

def sorted_keys(by, keys)
  x = {}

  case by
  when 'hour'
    keys.each do |k|
      y = format_date(k)
      next unless y

      x[y] = k
    end
  when 'day'
    keys.each do |k|
      y = format_day(k)
      next unless y

      x[y] = k
    end
  when 'ip'
    keys.each do |k|
      x[k] = k
    end
  when 'path'
    keys.each do |k|
      x[k] = k
    end
  end

  x
end

options, arguments = opts(%w[--csv])

valid_options(options)

report = options[REPORT]
by     = options[BY]

data = {}

if arguments.any?
  arguments.each do |filename|
    fh = File.open(filename, 'r')
    process_line_from_file(fh, data, by, REPORT_VALUES[report])
    fh.close
  end
else
  process_line_from_file(STDIN, data, by, REPORT_VALUES[report])
end

header, line = setup_display(by, report, data)

if options[CSV]
  puts header.map { |h| h.strip}.join(',')

  sk = sorted_keys(by, data.keys)

  sk.keys.sort.each do |k|
    v = data[sk[k]]

    case report
    when 'status'
      puts [k, v.count, v.code2, v.code3, v.code4, v.code5].join(',')
    when 'size'
      puts [k, v.count, v.total, v.min, v.average, v.max].join(',')
    when 'response'
      puts [k, v.count, v.min, v.average, v.max].join(',')
    end
  end
else
  puts "| #{header.join(' | ')} |"
  puts "+-#{header.map { |i| '-' * i.size }.join('-+-')}-+"

  x = "| #{line.join(' | ')} |"

  sk = sorted_keys(by, data.keys)

  sk.keys.sort.each do |k|
    v = data[sk[k]]

    case report
    when 'status'
      puts x % [k, v.count, v.code2, v.code3, v.code4, v.code5]
    when 'size'
      puts x % [k, v.count, v.total, v.min, v.average, v.max]
    when 'response'
      puts x % [k, v.count, v.min, v.average, v.max]
    end
  end
end

File: ngxl.go
package main

import (
	"bufio"
	"errors"
	"flag"
	"fmt"
	"github.com/PeterHickman/toolbox"
	strftime "github.com/ncruces/go-strftime"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

type stats struct {
	count         int64
	r_sec_min     float64
	r_sec_max     float64
	r_sec_total   float64
	r_bytes_min   int64
	r_bytes_max   int64
	r_bytes_total int64
	r_code_2xx    int64
	r_code_3xx    int64
	r_code_4xx    int64
	r_code_5xx    int64
}

const by_hour string = "%Y-%m-%d %H"
const by_day string = "%Y-%m-%d"
const by_ip string = "ip"
const by_path string = "path"

var time_format string
var group_format string
var what_to_report string
var key_width string
var key_title string
var key_bar string
var csv bool

func (s *stats) update(status string, bytes int64, seconds float64) {
	s.count += 1

	if s.count == 1 {
		s.r_bytes_min = bytes
		s.r_bytes_max = bytes
		s.r_bytes_total = bytes

		s.r_sec_min = seconds
		s.r_sec_max = seconds
		s.r_sec_total = seconds
	} else {
		if bytes < s.r_bytes_min {
			s.r_bytes_min = bytes
		} else if bytes > s.r_bytes_max {
			s.r_bytes_max = bytes
		}
		s.r_bytes_total += bytes

		if seconds < s.r_sec_min {
			s.r_sec_min = seconds
		} else if seconds > s.r_sec_max {
			s.r_sec_max = seconds
		}
		s.r_sec_total += seconds
	}

	if status >= "500" {
		s.r_code_5xx += 1
	} else if status >= "400" {
		s.r_code_4xx += 1
	} else if status >= "300" {
		s.r_code_3xx += 1
	} else {
		s.r_code_2xx += 1
	}
}

func reformat(line string) []string {
	parts := strings.Fields(line)

	output := []string{}
	current_token := ""

	for _, v := range parts {
		if strings.HasPrefix(v, "\"") {
			if current_token != "" {
				output = append(output, current_token)
			}
			current_token = v
		} else if strings.HasSuffix(v, "\"") {
			current_token += fmt.Sprintf(" %s", v)
			output = append(output, current_token)
			current_token = ""
		} else if strings.HasPrefix(v, "[") {
			if current_token != "" {
				output = append(output, current_token)
			}
			current_token = v
		} else if strings.HasSuffix(v, "]") {
			current_token += fmt.Sprintf(" %s", v)
			output = append(output, current_token)
			current_token = ""
		} else {
			if strings.HasPrefix(current_token, "\"") {
				current_token += fmt.Sprintf(" %s", v)
			} else if strings.HasPrefix(current_token, "[") {
				current_token += " "
				current_token += v
			} else {
				output = append(output, v)
			}
		}

	}

	if current_token != "" {
		output = append(output, current_token)
	}

	return output
}

func http_offset(parts []string) int {
	for i, v := range parts {
		if strings.Contains(v, " HTTP") {
			return i
		}
	}

	return -1
}

func extract_date(parts []string) (string, error) {
	for _, v := range parts {
		if strings.HasPrefix(v, "[") {
			x := strings.Fields(v)
			y, err := time.Parse(time_format, x[0][1:])
			if err != nil {
				return "", err
			}
			t := strftime.Format(group_format, y)
			return t, nil
		}
	}

	return "", errors.New("Unable to locate time stamp")
}

func extract_path(path string) string {
	parts := strings.Fields(path)

	i := strings.Index(parts[1], "?")

	if i == -1 {
		return parts[1]
	} else {
		return parts[1][:i]
	}
}

func init() {
	by := flag.String("by", "", "Group data by hour|day|ip")
	re := flag.String("report", "", "What information to report status|response|size")
	c := flag.Bool("csv", false, "Output as csv")

	flag.Parse()

	b := strings.ToLower(*by)

	if b == "" {
		usage("")
	}

	switch b {
	case "day":
		group_format = by_day
		key_width = "-10"
		key_title = "day"
		key_bar = "----------"
	case "hour":
		group_format = by_hour
		key_width = "-13"
		key_title = "date_and_hour"
		key_bar = "-------------"
	case "ip":
		group_format = by_ip
		key_width = "-15"
		key_title = "ip_address"
		key_bar = "---------------"
	case "path":
		group_format = by_path
		key_width = "-15"
		key_title = "path"
		key_bar = "---------------"
	default:
		usage(fmt.Sprintf("Unrecognised --by option %s", b))
	}

	time_format, _ = strftime.Layout("%d/%b/%Y:%H:%M:%S")

	r := strings.ToLower(*re)

	if r == "" {
		usage("")
	}

	switch r {
	case "status":
		what_to_report = r
	case "response":
		what_to_report = r
	case "size":
		what_to_report = r
	default:
		usage(fmt.Sprintf("Unrecognised --report option %s", r))
	}

	csv = *c
}

func usage(message string) {
	if message != "" {
		fmt.Printf("ERROR: %s\n\n", message)
	}

	fmt.Println("ngxl --report status|size|reponse --by hour|day|ip|path [--csv] <list of Nginx log files>")
	fmt.Println()
	fmt.Println("  Will process the nginx log files and report either the status (by class),")
	fmt.Println("  size of the response in bytes (giving the total, minimum, average and maximum)")
	fmt.Println("  or response times (also with minimum, average and maximum)")
	fmt.Println()
	fmt.Println("  When grouping by path the query portion will be ignored, /fred?a=1 becomes /fred")
	fmt.Println()
	fmt.Println("  Each line of the report will be either the day or hour that the data was")
	fmt.Println("  recorded in or the ip address that the request was made from")
	fmt.Println()
	fmt.Println("  The log lines can be read from stdin if you need to pipe them (through")
	fmt.Println("  grep perhaps)")
	fmt.Println()
	fmt.Println("  The output will be formatted as CSV data if the --csv flag is given")

	os.Exit(1)
}

func process_contents(data map[string]stats, file *os.File) {
	scanner := bufio.NewScanner(file)

	var d string
	var e error

	for scanner.Scan() {
		line := scanner.Text()
		parts := reformat(line)
		http := http_offset(parts)

		if http != -1 {
			if group_format == by_ip {
				d = parts[0]
				e = nil
			} else if group_format == by_path {
				d = extract_path(parts[http])
				e = nil
			} else {
				d, e = extract_date(parts)
			}

			if e == nil {
				val, ok := data[d]

				if !ok {
					data[d] = stats{}
				}

				// Status code for the response
				status := parts[http+1]

				// Size of response in bytes
				bytes, _ := strconv.ParseInt(parts[http+2], 10, 64)

				// Response time in seconds
				seconds, _ := strconv.ParseFloat(parts[http+3], 64)

				val.update(status, bytes, seconds)
				data[d] = val
			}
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
}

func process_file(data map[string]stats, filename string) {
	if !toolbox.FileExists(filename) {
		log.Fatalf("Cannot find logfile %s", filename)
	}

	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}

	process_contents(data, file)

	file.Close()
}

func report(data map[string]stats) {
	keys := make([]string, 0, len(data))

	max_width := 0

	for k := range data {
		l := len(k)

		if l > max_width {
			max_width = l
		}

		keys = append(keys, k)
	}

	if group_format == by_path {
		key_width = fmt.Sprintf("-%d", max_width)
		key_bar = strings.Repeat("-", max_width)
	}

	sort.Strings(keys)

	var format string

	switch what_to_report {
	case "response":
		fmt.Printf("| %"+key_width+"s |    count |      min |      avg |      max |\n", key_title)
		fmt.Println("+-" + key_bar + "-+----------+----------+----------+----------+")
		format = "| %" + key_width + "s | %8d | %8.3f | %8.3f | %8.3f |\n"
	case "status":
		fmt.Printf("| %"+key_width+"s |    count |      2xx |      3xx |      4xx |      5xx |\n", key_title)
		fmt.Println("+-" + key_bar + "-+----------+----------+----------+----------+----------+")
		format = "| %" + key_width + "s | %8d | %8d | %8d | %8d | %8d |\n"
	case "size":
		fmt.Printf("| %"+key_width+"s |    count |           total |             min |             avg |             max |\n", key_title)
		fmt.Println("+-" + key_bar + "-+----------+-----------------+-----------------+-----------------+-----------------+")
		format = "| %" + key_width + "s | %8d | %15d | %15d | %15d | %15d |\n"
	}

	for _, k := range keys {
		v := data[k]

		switch what_to_report {
		case "response":
			fmt.Printf(format, k, v.count, v.r_sec_min, (v.r_sec_total / float64(v.count)), v.r_sec_max)
		case "status":
			fmt.Printf(format, k, v.count, v.r_code_2xx, v.r_code_3xx, v.r_code_4xx, v.r_code_5xx)
		case "size":
			fmt.Printf(format, k, v.count, v.r_bytes_total, v.r_bytes_min, v.r_bytes_total/v.count, v.r_bytes_max)
		}
	}
}

func report_csv(data map[string]stats) {
	keys := make([]string, 0, len(data))

	for k := range data {
		keys = append(keys, k)
	}

	sort.Strings(keys)

	switch what_to_report {
	case "response":
		fmt.Printf("%s,%s,%s,%s,%s\n", key_title, "count", "min", "avg", "max")
	case "status":
		fmt.Printf("%s,%s,%s,%s,%s,%s\n", key_title, "count", "2xx", "3xx", "4xx", "5xx")
	case "size":
		fmt.Printf("%s,%s,%s,%s,%s,%s\n", key_title, "count", "total", "min", "avg", "max")
	}

	for _, k := range keys {
		v := data[k]
		switch what_to_report {
		case "response":
			fmt.Printf("%s,%d,%.3f,%.3f,%.3f\n", k, v.count, v.r_sec_min, (v.r_sec_total / float64(v.count)), v.r_sec_max)
		case "status":
			fmt.Printf("%s,%d,%d,%d,%d,%d\n", k, v.count, v.r_code_2xx, v.r_code_3xx, v.r_code_4xx, v.r_code_5xx)
		case "size":
			fmt.Printf("%s,%d,%d,%d,%d,%d\n", k, v.count, v.r_bytes_total, v.r_bytes_min, v.r_bytes_total/v.count, v.r_bytes_max)
		}
	}
}

func main() {
	data := map[string]stats{}

	if len(flag.Args()) > 0 {
		for _, filename := range flag.Args() {
			process_file(data, filename)
		}
	} else {
		process_contents(data, os.Stdin)
	}

	if csv {
		report_csv(data)
	} else {
		report(data)
	}
}
